torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -525,32 +525,43 @@ def forward(self, hidden_states):
         # for each token, select top-k experts, and compute the weight for each expert
         topk_idx, topk_weight = self.gate(hidden_states)
         hidden_states = hidden_states.view(-1, hidden_states.shape[-1])
-        y = self.moe_infer(hidden_states, topk_idx, topk_weight).view(*orig_shape)
+        y = self.moe_forward(hidden_states, topk_idx, topk_weight).view(*orig_shape)
         if self.config.n_shared_experts is not None:
             y = y + self.shared_experts(identity)
         return y
 
-    @torch.no_grad()
-    def moe_infer(self, x, topk_ids, topk_weight):
+    def moe_forward(self, x, topk_ids, topk_weight):
         if self.ep_size > 1 and (not self.has_symm_mem):
             # Set up symmetric memory for the first time, then reuse it
             self.setup_symm_mem(x.dtype, x.device)
 
-        # [seq_len, n_routed_experts]
-        cnts = topk_ids.new_zeros((topk_ids.shape[0], len(self.experts)))
-        # Fill 1 to the selected experts
-        cnts.scatter_(1, topk_ids, 1)
-        tokens_per_expert = cnts.sum(dim=0)
-        # Token indices for each expert
-        idxs = topk_ids.view(-1).argsort()
-        sorted_tokens_shape = idxs.shape + x.shape[1:]
+        # This part sorts the token indices so that tokens routed to the same expert reside consecutively.
+        # An implication is that tokens to the same "expert group" (i.e., device) are also consecutive.
+        # Since this is an "aritificial" index creation (final outcome being
+        # `idxs`), we don't need gradients here.
+        with torch.no_grad():
+            # [seq_len, n_routed_experts]
+            cnts = topk_ids.new_zeros((topk_ids.shape[0], len(self.experts)))
+            # Fill 1 to the selected experts
+            cnts.scatter_(1, topk_ids, 1)
+            tokens_per_expert = cnts.sum(dim=0)
+            # Token indices for each expert
+            idxs = topk_ids.view(-1).argsort()
+            sorted_tokens_shape = idxs.shape + x.shape[1:]
+
         # Take necessary space from the `token_send_buf` symm mem
         sorted_tokens = self.token_send_buf[: idxs.shape[0]]
+        sorted_tokens.copy_(x[idxs // topk_ids.shape[1]])
```

Comment:
x: [b*s, h]?
idxs: [b*s]?
topk_ids: [s, k]

idxs represents "the indices of tokens respective to b*s flattening, such that the tokens for expert0 come first, etc", right?

i'm confused about what `idxs // topk_ids.shape[1]` is doing.

Reply:
ok- the explanation is
1) idxs contains the indexes of tokens in a sequence that is K times longer than the original input sequence, due to the view(-1) above before the argsort
2) the index values logically refer to tokens in the original input sequence (x) so to make them useful again we floor-didide by `k` (the value used for topk).